### ISSUE TITLE
Font issue: findfonts should differentiate between thin and regular ttf fonts

### ISSUE THREAD

**Issue by blatts** [matplotlib/matplotlib#8788]
### Bug report

**Bug summary**

The selection criteria for finding fonts using the algorithm in ``matplotlib.font_manager.findfonts`` are incomplete. The problem arises with high-quality TTF fonts that have many different styles, for instance "Helvetica Neue", or "Lato". For instance, the "Thin" versions of these fonts are indistinguishable from the "Regular" versions by their font weight (both have e.g. a value of 400) such that the font that is returned by findfonts depends on the ordering of the list ``matplotlib.font_manager.fontManager.ttflist``.

A workaround would be to simply uninstall the unwanted versions of the conflicting fonts from the system. Unfortunately, for the figures that I am trying to create, "Helvetica Neue" is mandatory and it is also an OSX system font, thus cannot be disabled.

I am unsure whether the fonts (for instance) "HelveticaNeue-Thin.ttf" and "HelveticaNeue.ttf" could be distinguished by adding another font metric to the ``Font`` class. If not, the proposed solution would be to allow selecting fonts in ``matplotlib.rcParams`` by their file name, e.g. setting ``mathtext.rm : "HelveticaNeue.ttf"``. This solution would require changing the fontconfig parser in ``matplotlib.fontconfig_pattern.py`` to allow filenames.

**Code for reproduction**

The result of the code below is system-dependent, even when exactly the same font files are installed, because of the ordering of ``matplotlib.font_manager.fontManager.ttflist``.

```python
import matplotlib.font_manager as fm
p = fm.FontProperties(family = 'Helvetica Neue')
fm.findfont(p)

--> 'HelveticaNeue.ttf' or 'HelveticaNeue-Thin.ttf'
```

The properties of the corresponding Font objects are identical (see ``matplotlib.font_manager.fontManager.ttflist``)

```python
<Font 'Helvetica Neue' (HelveticaNeue.ttf) normal normal 400 normal>
<Font 'Helvetica Neue' (HelveticaNeue-Thin.ttf) normal normal 400 normal>
```

**Actual outcome**

See above, system-dependent.

**Expected outcome**

Findfonts should be deterministic.

**Matplotlib version**
  * Operating System: OSX, Windows
  * Matplotlib Version: 2.0.0
  * Python Version: 3.6.1

**Comment by tacaswell** [comment]
Related to #8551 #8607 #5250 #8550 #7931 

This seems to have become a major problem recently.

**Comment by lukelbd** [comment]
Hope this can be addressed soon. This has been a recurring issue when [adding new fonts](https://proplot.readthedocs.io/en/latest/colors_fonts.html#Included-font-names) to proplot -- I have to manually delete all of the "thin" files.

Looks like the issue is in `font_manager.ttfFontProperty`. Here is where styles and weights are read:

https://github.com/matplotlib/matplotlib/blob/5fb9709128feb71e2cbaa20f625c23c96d6a2ba1/lib/matplotlib/font_manager.py#L358-L438

What can we do to detect the `'thin'` property? Also is `'thin'` a style or a weight?

**Comment by anntzer** [comment]
If you're on linux (macos perhaps? if fontconfig is present) you can use https://github.com/anntzer/mplfccache to generate a more accurate font cache (asking fontconfig to list the font properties for you).

I have translated into Python the actual algorithm used by fontconfig to determine font properties at https://github.com/anntzer/freetypybind/blob/master/lib/freetypybind/query_face.py (using my own freetype binding but that shouldn't matter much); someone "just" needs to PR that into Matplotlib...

**Comment by anntzer** [comment]
Mostly closed by #16203.
